code/game/objects/structures/morgue.dm

```diff
@@ -242,67 +247,69 @@ GLOBAL_LIST_EMPTY(crematoriums)
 		else
 			src.icon_state = "crema1"
 
-		if(locked)
+		if(cremating)
 			src.icon_state = "crema_active"
 
 	return
 
-/obj/structure/bodycontainer/crematorium/proc/cremate(mob/user)
-	if(locked)
-		return //don't let you cremate something twice or w/e
-	// Make sure we don't delete the actual morgue and its tray
-	var/list/conts = GetAllContents() - src - connected
+/obj/structure/bodycontainer/crematorium/process()
+	if(cremating)
+		cremate()
 
-	if(!conts.len)
-		audible_message("<span class='hear'>You hear a hollow crackle.</span>")
-		return
-
-	else
-		audible_message("<span class='hear'>You hear a roar as the crematorium activates.</span>")
+///Toggles cremation. The cremate parameter can be used to set a specific value instead of toggling.
+/obj/structure/bodycontainer/crematorium/proc/toggle_cremation(set_cremate, mob/user)
+	if(cremating && set_cremate != TRUE)
+		cremating = FALSE
+		locked = FALSE
+		STOP_PROCESSING(SSobj, src)
+		update_icon()
 
+		audible_message("<span class='hear'>You hear a roar as [src] activates.</span>")
+		return
+	if(!cremating && set_cremate != FALSE)
+		if(user)
+			last_user = user
+		else
+			last_user = null //let's not blame the previous guy
+		cremating = TRUE
+		first_cremation_tick = TRUE
 		locked = TRUE
+		START_PROCESSING(SSobj, src)
 		update_icon()
 
-		for(var/mob/living/M in conts)
-			if (M.stat != DEAD)
-				M.emote("scream")
-			if(user)
-				log_combat(user, M, "cremated")
-			else
-				M.log_message("was cremated", LOG_ATTACK)
-
-			M.death(1)
-			if(M) //some animals get automatically deleted on death.
-				M.ghostize()
-				qdel(M)
-
-		for(var/obj/O in conts) //conts defined above, ignores crematorium and tray
-			qdel(O)
-
-		if(!locate(/obj/effect/decal/cleanable/ash) in get_step(src, dir))//prevent pile-up
-			new/obj/effect/decal/cleanable/ash/crematorium(src)
-
-		sleep(30)
-
-		if(!QDELETED(src))
-			locked = FALSE
-			update_icon()
-			playsound(src.loc, 'sound/machines/ding.ogg', 50, TRUE) //you horrible people
-
-/obj/structure/bodycontainer/crematorium/creamatorium
-	name = "creamatorium"
-	desc = "A human incinerator. Works well during ice cream socials."
+		audible_message("<span class='hear'>The roar of [src]'s flames gradually dies down.</span>")
+		return
 
-/obj/structure/bodycontainer/crematorium/creamatorium/cremate(mob/user)
-	var/list/icecreams = new()
-	for(var/i_scream in GetAllContents(/mob/living))
-		var/obj/item/reagent_containers/food/snacks/icecream/IC = new()
-		IC.set_cone_type("waffle")
-		IC.add_mob_flavor(i_scream)
-		icecreams += IC
-	. = ..()
-	for(var/obj/IC in icecreams)
-		IC.forceMove(src)
+///Ignites and causes heavy burn damage to mobs and objects inside. Non-immune mobs will quickly be incinerated.
+/obj/structure/bodycontainer/crematorium/proc/cremate()
+	if(!cremating)
+		return

```

Comment:
you already checked this in process

Reply:
i opted to have more checks to be safe, but they can be removed if redundancy is a problem